### PR TITLE
Create converter to/from numerical and string properties #34

### PR DESCRIPTION
**Pull request by stefanhans**
Due to my sickness last week I had to finish it. It's not 100%, but it should work.

### CODE REVIEW

**Comment by stefanhans** [comment]
I guess the problem is `import _ "proofs/proto"` does not work on my environment. I was facing that problem right from the start and could not fix it yet.

**Comment by stefanhans** [comment]
`go test -race -coverprofile=coverage.txt -covermode=atomic -v ./converter`

works for me

```
...
PASS
coverage: 99.2% of statements
ok  	github.com/centrifuge/precise-proofs/converter	1.034s	coverage: 99.2% of statements
```

File: converter/converter.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
+package converter
+
```
Comment (lucasvo) [inline converter/converter.go]: Could you add a short description of the converter package here so it shows up in godoc?

File: converter/converter.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		if bp1[i] != bp2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+var prototypes = make(map[reflect.Type]*ProtoType)
+
+func GetBinaryProperty(messageTyp reflect.Type, literal string) ([]uint64, error) {
```
Comment (lucasvo) [inline converter/converter.go]: Please add descriptions for all public methods

File: examples/documents/example.proto [hunk 3]
```diff
@@ -185,3 +185,94 @@ message Integers {
   sfixed32 valueI = 9;
   sfixed64 valueJ = 10;
 }
+
+
+/*
+The following message types are for converter tests only
+*/
```
Comment (lucasvo) [inline examples/documents/example.proto]: Are the other messages completely unusable for the converter tests? I'd love to have not too many messages and reuse the same ones (also to make sure we test with the same data across everything)

File: converter/converter.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+		return protoType.handleStruct(typ)
+	case reflect.Map:
+		return protoType.handleMap(typ)
+	case reflect.Ptr:
+		// Resolve the pointer
+		return protoType.handleType(typ.Elem())
+	case reflect.Slice:
+		return protoType.handleSlice(typ)
+	case reflect.Uint32:
+		return nil
```
[the remark this answers is not in the document]
Comment (stefanhans) [reply converter/converter.go]: It seems that these reflect.Kind cannot be created by protoc. I've inserted first all reflect.Kind and deleted the one not matched by proto3 fields later:
```
type = "double" | "float" | "int32" | "int64" | "uint32" | "uint64"
      | "sint32" | "sint64" | "fixed32" | "fixed64" | "sfixed32" | "sfixed64"
      | "bool" | "string" | "bytes" | messageType | enumType
```

File: converter/converter.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+	// Set root message type as type if needed
+	if typ.Kind() != reflect.Slice {
+		return fmt.Errorf("%v is not a slice\n", typ)
+	}
+
+	if typ == reflect.SliceOf(reflect.TypeOf(uint8(0))) {
+		return nil
+	}
+
+	indexString := "[0]"
```
[the remark this answers is not in the document]
Comment (stefanhans) [reply converter/converter.go]: It seems we have a misunderstanding here. The type does not know about if there's more than one element in the map or repeated, right? So, `[0]` and `['']` just indicates that this type is a map or a slice (repeated). The next two comments are related.

File: converter/converter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucasvo) [reply converter/converter.go]: https://developers.google.com/protocol-buffers/docs/proto#scalar << this is a supported type though, right?

File: converter/converter.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucasvo) [reply converter/converter.go]: The type doesn't know the length but the behavior should still be that it it transforms that value. The same goes for maps. If you pass in a valid index (e.g. not a value that is too large for the type of index) then it should just convert it appropriately.

**Comment by stefanhans** [comment]
Today I have a little medical operation. I will be back tomorrow.

**Pull request closed without merging**